FEAT: +TOKEN modifier for side-effect functions

ankostis opened the pull request:
Inputs & outputs in ``needs`` & ``provides`` kwargs designated as ``token`` work as usual while solving the DAG but they are never assigned any values to/from the operation's functions:
```python
from graphkit import operation, compose
from graphkit.modifiers import token

# Function appending a new dataframe column from two pre-existing ones.
def addcolumns(df):
    df['sum'] = df['a'] + df['b']

# Designate `a`, `b` & `sum` column names as an token arguments.
graph = compose('mygraph')(
    operation(
        name='addcolumns',
        needs=['df', token('a'), token('b')],
        provides=[token('sum')])(addcolumns)
)

# The graph works with and without 'c' provided as input.
df = pd.DataFrame({'a': [5], 'b': [2]})
assert graph({'df': df})['sum'] == 11
```

---
I confirm that this contribution is made under the terms of the license found in the root directory of this repository's source tree and that I have the authority necessary to make this contribution on behalf of its copyright owner.

ankostis committed:
FEAT: +TOKEN modifier for side-effect functions

Tokens work as usual while solving the DAG but
they are never assigned any values to/from the operation's functions.

+ TC included.
+ Docs updated.\+ Added `modifiers` superclass to facilitate identification code.
+ refactored FunctionalOperation._compute().

ankostis committed:
fix(func): 1-item outs broke with token modifiers...

+ _norm_kwargs:
  + use isinstance() instead of type-equality checks,
    to support new modier classes;
    + avoid excessive dict searches with local vars.

yahoocla commented:
Thank you for submitting this pull request, however I do not see a valid CLA on file for you. Before we can merge this request please visit https://yahoocla.herokuapp.com/ and agree to the terms. Thanks! :smile:

ankostis commented:
Just didi it.

Is it understood what I'm trying to achieve?

ankostis committed:
FIX(build): py2 needs pinning networkx-2.2

ankostis committed:
refact(token): undo adding of new `modifier` class...

not really needed, better be explicit which modifier is searched.

ankostis committed:
enh(token.TC): test 2 reversed token workflows

ankostis committed:
ENH(net,#18): ignore UN-SATISFIABLE operations with partial inputs

+ The x2 TCs added just before are now passing.

ankostis committed:
doc(#13): sample code to plot workflow diagram in intro

ankostis committed:
FIX(#13): BUG in plot-diagram writtin from PY2 era,

were writing in text-mode in PY3. and failing as encoding error.

ankostis committed:
enh(plot): provide help msg on supported file-exts

ankostis committed:
TEST(plot,ci): test plotting; pip install extras in Travis

ankostis committed:
fix(plot): don't create file on unsupported formats

thanks to @andres-fr.

ankostis committed:
enh(plot.TC): expose supported writers and TC on them

ankostis committed:
ENH(OPTIONAL): optionals per OPERATION, not to NEEDs...
o
+ Closes #17

ankostis committed:
enh(test): + x2 TC breaking UNSATISFIED operations...

receiving partial inputs, needed for other operations.

ankostis committed:
refact(net): drop old `dag` nx-package

ankostis committed:
ENH(core): ORDERED SETs for DETERMINISTIC results

NOTE dict are not deterministic in <PY3.6.
So this commit would not improve determinism
in those pythons.

+ build: add `boltons` dependency for ndexedSet.
+ doc: mark all set usage if affect determinism.
+ e.g. see reproducibility problem in #14.

ankostis committed:
merge UNASTIFIABLE + ORDERED_SETs

ankostis committed:
refact(net): simpilify del-instruction loop

ankostis committed:
REFACT(unsatisfied): doubly-recursive func --> loop on topo-sorted

ankostis committed:
test(dag,#25): FAILing TC for overriding intermediate data

ankostis committed:
test(dag,#24): FAILing TC for over-pruning inetermediates when outs asked

ankostis committed:
MERGE (prune_unsatified, ordered_sets) into fix-pruning ...

needed to refactor the new pruning algorithm.

- expected 2 TCs fail for yet-to-be-solved #24 & #25 bugs.

ankostis committed:
DOC(net): explain DAG solution & compilation...

the later described in #21.

ankostis committed:
TEST(prune): +Failing x2 TCs multi-out must run but not...

override intermediate data.

More changes only for newer pruning TCs:
+ refact(test): rename graph-->netop vars for results of compose(),
  to avoid of `graph.net.graph`.
+ Explain failure modes in v1.2.4 & this merged branch (#19 + #23).

ankostis committed:
refact(dag): call compile() before compute.compute...

not after compose().

+ All TCs pass ok.
+ NOTE this is not yet what is described in #21.

ankostis committed:
test(dag): +TC checking DeleteInst vary when inputs change

ankostis committed:
doc(net): explain new DAG SOLUTION

ankostis committed:
REFACT(NET) COMPILE+COMPUTE...

+ Read the next doc-only commit to understand changes.
+ Renamed:
  + net.steps --> net.execution_plan.
  + (old)compile() --> _build_execution_plan()
  + _find_necessary_steps() --> (new)compile() + _solve_dag()
    compile() became the master function invoking _solve_dag &
    _build-execution_plan(), and do the caching.
+ refact(compute()): extract common tasks from sequential/parallel.
+ refact show_layers() to allow full-print, geting also string
  (not just printing), and using custom classes for representation.
+ Raise AssertionError when invalid class in plan.
  it's a logic error, not a language type-error.

ankostis committed:
ENH(net): move compile() after SOLVE DAG ...

to pass +TC checking DeleteInst vary when inputs change.

- x4 TCs still failing, and need revamp of dag-solution.

ankostis committed:
FIX(net): new Ops invalidate execution-plan cache...

Probaly unreported bug in v1.2.4 for '_neccessary_steps_cache`.

ankostis committed:
ENH(DAG): NEW SOLVER

+ Pruning behaves correctly also when outputs given;
  this happens by breaking incoming provide-links
  to any given intermedediate inputs.
+ Unsatisfied detection now includes those without outputs
  due to broken links (above).
+ Remove some uneeded "glue" from unsatisfied-detection code,
  leftover from previous compile() refactoring.
+ Renamed satisfiable --> satisfied.
+ Improved unknown output requested raise-message.
+ x3 TCs PASS, x1 in #24 and the first x2 in #25.
- 1x TCs in #25 still FAIL, and need "Pinning" of given-inputs
  (the operation MUST and MUST NOT run in these cases).

ankostis committed:
WIP/FIX(prune,#26): PIN intermediate inputs if operation before must run

- WIP: PARALLEL execution not adding PINS!
+ Insert "PinInstructions" in the execution-plan to avoid overwritting.
+ Add `_overwrite_collector` in `compose()` to collect re-calculated values.
+ FIX the last TC in #25.

ankostis committed:
enh(build): replace numpy with pytest...

numpy was used just for its assert_raise

ankostis committed:
feat(build): add pip-extras [test]

ankostis committed:
enh(build,ci): use pytest in travis

ankostis committed:
REFACT(net): part 3 of new dag-solver & pin refactoring

- STILL buggy PIN on PARALLEL, 2 DISABLED TCs FAIL:
  - test_pruning_with_given_intermediate_and_asked_out()
  - test_unsatisfied_operations_same_out()
+ move check if value in asked outputs before cache-evicting it
  in build-execution-plan time - compute methods
  don't need outputs anymore.
+ test: speed up parallel/multihtread TCs
  by reducing delays & repetitions.
+ refact: network rightfully adopted stray functions
  for parallel processing - they all worke on the net.graph,
+ upd: networkx api by indexing on `dag.nodes` views.
+ enh: add log message when deleting in parallel
  (in par with sequential code).
+ refact: var-renames, if-then-else simplifications, pythonisms.
+ doc: A lot!

ankostis committed:
WIP/FIX(PIN): PARALLEL DELs decide on PRUNED-dag (not full)...

- WIP: x4 TCs FAIL and still not discovered th bug :-(
+ BUT ALL+AUGMENTED PARALLEL TCs pass
  (#26 were failing some)
+ refact: net stores also `pruned_dag` (not only `steps`).
+ refact: _solve_dag() --> _prune_dag().
+ doc: +a lot.
+ TODO: store pruned_dag in own ExePlan class.

ankostis committed:
WIP/+ExecPlan class...

- SAME x4 TCs FAIL like parent.
+ refact: revive net.steps --> net.last_plan
+ BREAK: inverse the order of outputs/inputs args in net.compute()
+ refact: use iset() to merge.

ankostis committed:
enh(CI): +PY3.6 where dicts are stable

ankostis committed:
fix(plot): NetOp did not return pydot instance

ankostis committed:
refact(plot): extract plot function out of Network class...

to use it also on stary DAGs.  Keep delegation.

ankostis committed:
fix(plot): matplotlib plot was failing  in PY3 due IO io misuse

ankostis committed:
FEAT(plot): overlay Execution STEPS on diagrams

ankostis committed:
ENH(plot): +inputs, +outputs, +solution modify plotting

(see #13 for an example):

ankostis committed:
refact(plot.TC): move plot tests to early beggining

ankostis committed:
doc(plot): tell supported formats in doctest, +TC

ankostis committed:
doc(plot): add legend & example; docstring in netop.plot()

ankostis committed:
enh(plot): mark optional "needs"

ankostis committed:
ENH(plot): visual enhamcents on nodes & edges

ankostis committed:
test(plot): enhance plot test to try all #13 features;

+ test all chained plot() methods from netop.

ankostis committed:
test(optional): +x1 TC to check selective optionals with same out...

+ enh TC with parallel.
+ Scavenged from #20.

ankostis committed:
doc(plot); explain also params in user-facing API

ankostis committed:
FIX(PLOT.TC): TC was always testing PNG, ...

+ retorfitted to try all available formats.
+ list of forbidden formats based on my failres

ankostis committed:
fix(plot): don't require Matplotlib if no Window asked

ankostis committed:
test(plot): check also matplotlib show=True

ankostis committed:
ENH(plot): return SVG rendered in JUPYTER, ...

+ doc: rename in sample code: netop --> pipeline.
+ enh(build): add `ipython` in test dependencies.
+ include it in the plot TC.

ankostis committed:
refact(plot.TC): avoid writting multiple temp-files

ankostis committed:
MERGE PYTEST into PLOT to parametrize TCs.

ankostis committed:
fix(build): reuse dependencies definitions

ankostis committed:
REFACT(plot.TC): PYTESTize and parametrize

ankostis committed:
REFACT(PLOT): MOVE PLOT in own module

ankostis committed:
DROP PY3.4 - add PY3.6, PY3.7...

...pytest has problems with 3.4.

ankostis committed:
refact(plot): separate graphviz building from IO

ankostis committed:
MERGE ENHANCED-PLOTS into REFACT-ExecutionPlan to debug issues

ankostis committed:
DROP(net): list/show layers not needed, repr() is ok

ankostis committed:
ENH(plot,net): +plot() on ExecPlan; +PlotMixin ...

to avoid copying plot() doc+sig around

ankostis committed:
enh(plan): repr()

ankostis committed:
refact(plot): reorder formats-list with io

ankostis committed:
FIX(plot): distinguish Del/Pin cmds; +choice utils

- still x4 TCs fail...

ankostis committed:
refact(plot): inline imports, not to cycle with base...

to bring back PlotMixin to change together stuff.

ankostis committed:
refact(plot): move PlotMixin base-->plot module to group edit

ankostis committed:
enh(plot): +`title` arg at the bottom

ankostis committed:
FIX(plot): failing if steps not a list/ is none

ankostis committed:
ENH(plan,plot): executed operations drawn as filled

+ enh(plan): also sequential execution collects accounts executed.
+ refact(plan): executed_nodes-->executed.

ankostis committed:
FIX(net): revived last_plan was never set (HEAD~13: 64e0028)

ankostis committed:
refact(plot): reodred build-dot utils above use

ankostis committed:
doc(plot): move centrally all API doc on PlotMixin; Style

ankostis committed:
FEAT(plot,plan): +yellow broken links on original graph

ankostis committed:
refact(plot): PlotMixin --> Plotter, _plotter() --> _plot()

ankostis committed:
TEST(PLOT): Check also ExecPlan

ankostis committed:
REFACT(PLOT): Plotter builds dot & renders...

... based on Law of Demeter simplify kwargs,
defined in one place.

+ enh: plotters "suggest" kwargs, possinly to override them
  (no duplcate arg when attempted).

ankostis committed:
ENH(plan.polt): CLUSTER pruned nodes

ankostis committed:
FIX(plot): VISUAL fixes & COLOR-palettes:

+ FIX: apply thickness as in legend also for operators.
+ enh: don't cluster if no nodes pruned.
+ enh: netop includes its name a graph-title.
+ color palette: wheat filled nodes.
+ reuse common func.

ankostis committed:
ENH(netop): mark all its needs as OPTIONAL

ankostis committed:
FIX(NET): were FORGETTING PRUNED ASKED-OUTPUTs...

... bugged in the opening commit d403783 of this PR, and
discovered 68(!) commits later, and all that time had to live
with x4 broken TCs with asked-outputs.

ankostis committed:
ENH(plot,TC): pipeline delegates to last_plan, if exists

ankostis committed:
FIX(plot.TC): formats-TC were testing dot-file, not rendered...

+ ENH: strongly refatctored TC to detect nulls & dupes.

ankostis committed:
enh(net): prune also isolate data

ankostis committed:
FIX(<PY3.5): ORDERED DiGRAPH for old Python to fix TCs

+ Partial fix deterministic results (#22-2.4.3i).

ankostis committed:
ENH(plot): switch oval<-->circle ops/netops

ankostis committed:
feat(plot): +legend & TC

ankostis committed:
DOC(PLOT): +sample GRAPH & LEGEND into README...

+ test: some refacts on vars & for the uploaded images.
+ MARK SPURIOUS FAILS in < PY3.6 due to unordered dicts
  eg https://travis-ci.org/ankostis/graphkit/jobs/594813119

ankostis committed:
DROP(plot) jupyter kw, simpler to monkeypath Dot class...

to support _repr_svg(), and it works without any flag.

ankostis committed:
Rebrand TOKENS --> SIDEFFECTS

ankostis committed:
FIX(sideffect.TX): old TC was not with unsatisfied in mind,

needed to give the sidefeect as input.

ankostis committed:
FIX(plot): LEGEND mistakes, SVGize, egg graphs, ...

+ graphop label renames,
+ DOT shortening.

ankostis committed:
FIX(DOC): don't use `graph` in sample code, crash DOT, +more ...

+ FIX: `graph` is a DOT keyword, eg  pydot/pydot#111
  + Replaced `graph` --> `graphop`, inline with future refactoring.
+ Refactored example code.
+ New "Plotting" section under Quickstart.
+ Copy README instructions to main docs.
+ Abandoned `example_graph.svg`, replaced with `intro.svg`.
+ Recommend `.png` --> `.svg`, to save storage space.
- discovered BUG in MERGE saumple code (empty!).

ankostis committed:
chore(git): .gitignore plot images in root folder

ankostis committed:
chore(TCs):  mark SLOW tests, when in hurry, +`setup.cfg`

... to adopt pytest args
+ mark wheel a universal.

ankostis committed:
FIX(TCs): MERGE TCs were not ASSERTING...

got values from v1.2.4
- ALL MERGE TCs FAIL!

ankostis committed:
FIX(MERGE): broken by NEW_DAG_SOLVER (#26 ...

many commits ago.

Never got it bc TC were not checking merges!

ankostis committed:
MERGE NEW_SOLVER + PLOT --> SIDEFFECTS + OPTS in DAG
